Use transaction in attrd_write_attribute(), deprecate cib_mixed_update

nrwahl2 opened the pull request:
I need to re-run cts-lab on this. It was okay as of a few weeks ago.

kgaillot reviewed (commented):
Didn't get to last two commits yet, but looking good so far

nrwahl2 commented on include/crm/cib/internal.h:
```diff
@@ -164,6 +164,39 @@ struct timer_rec_s {
+ * \return \c true if the CIB client name is trusted, or \c false otherwise
+ */
+static inline bool
+cib__client_name_is_trusted(const char *name)
```
Initially I used "safe" and changed it to "trusted" last night (we trust it, but we can never guarantee anything is safe haha). How do you feel about "safe"?

kgaillot replied:
safe would be fine, but how about something like `cib__should_ignore_client()`?

nrwahl2 replied:
We're not ignoring it in general, only in the context of daemons refreshing the CIB. Sigh, naming things is hard :)

nrwahl2 replied:
On that note, I'm thinking `cib__client_triggers_refresh()` and reversing the logic.

nrwahl2 commented:
Only changes are `s/trusted/safe/g`

nrwahl2 commented on daemons/attrd/attrd_cib.c:
```diff
@@ -485,12 +566,6 @@ attrd_write_attribute(attribute_t *a, bool ignore_delay)
         if (v->current) {
             v->requested = strdup(v->current);
-        } else {
-            /* Older attrd versions don't know about the cib_mixed_update
```
Looking at this anew, I'm wondering if we had an oversight re: transactions or if I'm forgetting something right now.

If we send a commit-transaction request in a mixed-version cluster, the older nodes may not recognize it... right?

It seems like that would apply to any new CIB request type that we've ever added though, and I don't remember seeing a lot of special handling. In particular, `cib_scope_local` does nothing unless we're in legacy mode.

---

Related: There are still some comments in the transaction code about syncing the result CIB. I don't believe we do that anymore, now that the whole transaction is built on the client side and bundled into one commit-transaction request. The commit-transaction request would just get sent to all nodes to process locally, and to activate on success.

nrwahl2 replied:
Oh wow, I don't think we've had any new CIB request types since 1.1.8 :/

kgaillot replied:
Bah, don't know how I overlooked that. We also have to consider a newer CIB client (command-line tool) on a remote node talking to an older cluster node, though maybe it's sufficient to just get a "not supported" error in that case.

My memory is hazy since it changed so much -- I thought the local CIB manager applied transactions to a copy of the CIB, then the result was sync'd to peers (including itself) via existing mechanisms.

kgaillot replied:
Currently attrd and execd implement their own protocol versions, and each daemon has its own IPC connection sequence. Eventually I'd like to get to where we have a common connection sequence (handshake) used with all daemons that would include passing key information like feature set and daemon protocol version supported. Not sure what based has currently.

nrwahl2 replied:
I think "not supported" would be fine for a remote client.

It did, in the original implementation where the whole transaction was built in the CIB manager. In the PR where we changed it to build the transaction in the client and send it as one big request with multiple nested requests, we also made it behave like any other "modifies" request. On v2 it gets sent to all nodes to apply locally (assuming I'm remembering the mechanism right) and on v1 it gets sent to the primary or broadcast or whatever.

nrwahl2 replied:
I don't think based has anything like a handshake that gets the feature set. It could be pulled from the CIB XML if that doesn't miss cases. My idea was to mimic the legacy mode determination. If we receive a request with feature set too low to support transactions, then we set a local_transactions boolean to true for as long as the based process remains active. If that's true, then any future transactions get processed locally and synced instead of sending it to all nodes to process. (Or do that based on the feature set in CIB XML.)

kgaillot replied:
If it's already a generic modify request by the time it's broadcast, then we don't need to worry about compatibility (other than remotes)

nrwahl2 replied:
It's never a modify request. It's a commit-transaction request, which has the modifies flag and contains an arbitrary sequence of modifying requests within it.

A backward compatible implementation would apply all changes locally and then sync it.

kgaillot replied:
In any case it should be able to operate that way -- the local daemon applies the transaction to a copy of the CIB, then generates a diff based on that and sends the diff around as if it were a single modify op

nrwahl2 replied:
> Commit is a local op, and IIRC the local daemon applies the entire transaction, and only the result is sent around. But I could be misremembering.

That was true originally. It changed when I realized we could put the whole transaction into an XML blob that doesn't need to maintain internal state in the local CIB manager.

nrwahl2 replied:
> In any case it should be able to operate that way -- the local daemon applies the transaction to a copy of the CIB, then generates a diff based on that and sends the diff around as if it were a single modify op

That could work -- a new `PCMK__CIB_REQUEST_APPLY_PATCH` request using a patchset generated from the local transaction result CIB. I don't know off hand how much easier or harder that would be than syncing the result CIB. It would probably have to copy or functionize what `sync_our_cib()` does to build a new request.

The only definite advantage would be less messaging overhead for large CIBs -- but that might be substantial.

---

Not sure if that should be THE way to implement transactions, or only compatibility code for when other nodes won't recognize "commit-transaction."

The reason we do it the current way (send the whole transaction request to all nodes to try to commit) is that that's how we treat every other request that modifies the CIB. It's not unreasonable to process a transaction locally and then apply the diff. I had just wanted to avoid special handling unless necessary.

kgaillot replied:
A replace instead of a diff would be fine. The goal would be to avoid adding any new ops that need to be sent around. If we can have a single way rather than new/legacy that would be better. Basically commit would be a local op, but it would generate a modify/replace/sync/whatever op that would be normal.

nrwahl2 replied:
Thanks. I'll look into both approaches and see which one seems cleaner.

By the way, the attrd changes in this PR are unlikely to change based on the changes to the transaction implementation (even though we can't merge this until the transaction implementation is fixed).

nrwahl2 replied:
Fixed via #3235

nrwahl2 commented on daemons/attrd/attrd_cib.c:
```diff
@@ -422,8 +473,29 @@ attrd_write_attribute(attribute_t *a, bool ignore_delay)
+            crm_err("Failed to write %s (id %s, set %s): Could not initiate "
+                    "CIB transaction",
+                    a->id, pcmk__s(a->uuid, "n/a"), pcmk__s(a->set_id, "n/a"));
+            return;
```
This might as well be a goto done too, for future-proofing

nrwahl2 commented:
Added the "NULL-check the_cib earlier" commit

nrwahl2 commented on daemons/attrd/attrd_cib.c:
```diff
@@ -301,48 +302,98 @@ attrd_cib_callback(xmlNode *msg, int call_id, int rc, xmlNode *output, void *use
+        rc = cib_internal_op(the_cib, PCMK__CIB_REQUEST_MODIFY, NULL,
+                             XML_CIB_TAG_STATUS, update, NULL,
+                             cib_can_create|cib_transaction, attr->user);
+        free_xml(update);
```
This needs to happen on failure too. `update` needs a broader scope so that we can free it in `done`.

nrwahl2 commented:
Most recent push only rebases on current main.

nrwahl2 commented:
Updated "transaction in attrd_write_attribute()" commit and added an "Avoid cib_internal_op()" commit on top

nrwahl2 commented:
s/cib__client_name_is_safe/cib__client_triggers_refresh/g and reversed the logic

nrwahl2 commented:
Last thing: I added the three commits from #3235 here, before "CIB transaction in attrd_write_attribute()". We can handle the other PR separately first, or handle those commits through this PR, but the transaction in attrd can't be merged until we transactions are backward compatible.

kgaillot commented on daemons/attrd/attrd_cib.c:
```diff
@@ -423,8 +522,29 @@ attrd_write_attribute(attribute_t *a, bool ignore_delay)
+            crm_err("Failed to write %s (id %s, set %s): Unknown set type %s",
+                    a->id, pcmk__s(a->uuid, "n/a"), pcmk__s(a->set_id, "n/a"),
+                    a->set_type);
+            goto done;
```
It occurs to me we shouldn't create attributes with invalid set types in the first place (in `attrd_create_attribute()`), so this would simplify. We could skip the check for private attributes, but all others should be rejected as soon as possible.

nrwahl2 replied:
You're proposing:
* In `attrd_create_attribute()`, if `PCMK__XA_ATTR_SET_TYPE` is invalid **and** `PCMK__XA_ATTR_IS_PRIVATE` is false (indicating we'll never write it to the CIB), return `NULL`
* Drop the set type validation from `attrd_write_attribute()`

Is that correct?

kgaillot replied:
Yes, though we have `a->is_private`

kgaillot reviewed (commented):
Looking good

nrwahl2 commented:
Rebased on current main

nrwahl2 commented on daemons/attrd/attrd_cib.c:
```diff
@@ -301,48 +302,146 @@ attrd_cib_callback(xmlNode *msg, int call_id, int rc, xmlNode *output, void *use
+ * \param[in] set_id     ID of attribute set
+ * \param[in] user       CIB user for update request
+ *
+ * \return Legacy Pacemaker return code
```
Good point

nrwahl2 commented on daemons/attrd/attrd_cib.c:
```diff
@@ -500,28 +623,33 @@ attrd_write_attribute(attribute_t *a, bool ignore_delay)
+        // Call cib_internal_op() directly to pass a->user
+        a->update = cib_internal_op(the_cib, PCMK__CIB_REQUEST_COMMIT_TRANSACT,
+                                    NULL, NULL, the_cib->transaction, NULL,
+                                    cib_none, a->user);
```
Thanks, this is a leftover from doing some things out of order.

nrwahl2 commented:
Updated `crm_is_daemon_name()` and `transaction in attrd_write_attribute()` commits to address review. Also added two commits on top to move set type validation to attribute creation time.

nrwahl2 commented:
Updated to fix set type default (NULL defaults to `XML_TAG_ATTR_SETS`)

kgaillot reviewed (approved):
Looks good, do you want to do any more testing?

nrwahl2 commented:
I think we're good to merge. 100 runs of cts-lab passed last night. I've only made trivial-looking changes today. I can kick off another cts-lab tonight. If anything unexpected happens I'll address ASAP.
